alpha/declcfg/diff.go

```diff
@@ -242,24 +242,27 @@ func bundlesEqual(b1, b2 *model.Bundle) (bool, error) {
 func addAllDependencies(newModel, oldModel, outputModel model.Model) error {
 	// Get every oldModel's bundle's dependencies, and their dependencies, etc. by BFS.
 	providingBundlesByPackage := map[string][]*model.Bundle{}
-	for curr := getBundles(outputModel); len(curr) != 0; {
-		reqGVKs, reqPkgs, err := findDependencies(curr)
+	var visitedBundles []*model.Bundle
```

Comment:
Should we store this as a map from the get-go, rather than regenerating the map during every call to `difference()`
```suggestion
	visitedBundles := map[*model.Bundle]struct{}{}
```

Reply:
Defining it as a slice here helps with working with the output of `getBundles` [here](https://github.com/operator-framework/operator-registry/pull/937/files#diff-d90d4db089c34d739cc336d4e85f9a924f2ce9c5ddda212f18f00515234513d6R247). The idea was to restrict the time complexity of difference to o(n), so doesn't defining `visitedBundle` as a map mean we won't essentially get any computational reduction, but instead move the complexity from one place to another?

Reply:
I suppose the difference would manifest if there was a way for lots of duplicates to be added to the `visitedBundles` list. But I didn't trace the code enough to know if that happens.

The scenario I'm thinking of:
- the catalog has 1000 bundles in it.
- `foo.v0.1.0` is a bundle in the catalog
- all other bundles depend on `foo.v0.1.0`.

In that scenario, does `visitedBundles` end up with 999 copies of `foo.v0.1.0`?

Reply:
In your example, with the 999 bundles existing in level 0 of the graph, `findDependencies` returning a map of the required gvks/packages ensures that `visitedBundles` has only one copy of `foo.v0.1.0`. 

It gets a little more interesting with a graph like this: 

(a, b, c...)-------------
|                             | ---- z
x------------------------

In that case, `visitedBundles` does end up with two copies of z. 

But, is preventing that worth converting `visitedBundles` map (with the 999 entires) to a list every time we enter the loop, i.e when we visit each level of the graph?

Reply:
Perhaps not. You've clearly thought about this more than I have so I'll definitely defer to you. And since this is all just discussion about optimization, I don't think its necessary to keep blocking on that.

If we later run into a case where this becomes a problem, we can take a deeper look then.